tests/test_ssh_keygen.py

```diff
@@ -0,0 +1,98 @@
+import io
+import subprocess
+import tempfile
+
+from dulwich import ssh_keygen
+
+from . import TestCase
+
+DOUBLE_CHECK_SSH_KEYGEN_SIMULATION = False
+
+
+msg_sig_pair_of_commit = (
+    b"""\
+tree fd7187a49a7b26eb36d782d34c672245b94b2e30
+parent b0162c139e4ea2e4783402de26617c912eef1e19
+author Castedo Ellerman <castedo@castedo.com> 1729263976 -0400
+committer Castedo Ellerman <castedo@castedo.com> 1729263976 -0400
+
+use vite-plugin-static-copy in doc example of Vite
+""",
+    b"""\
+-----BEGIN SSH SIGNATURE-----
+U1NIU0lHAAAAAQAAADMAAAALc3NoLWVkMjU1MTkAAAAghB1C63jrmh3eWRXJVbrTfw9wP/
+BIZf/aKPdFxBlMCq0AAAADZ2l0AAAAAAAAAAZzaGE1MTIAAABTAAAAC3NzaC1lZDI1NTE5
+AAAAQJoJUglNVFSaWhnbCl4WImCRLUEo6ymS9WBVOqpoH4kJVgIAMmCIAq9yDOL4fGYmJ0
+GsF7btQ1wFf6sbMzq9nwA=
+-----END SSH SIGNATURE-----
+""",
+)
+
+### The following test case produced by
+### echo Hola Mundo > hola.txt
+### ssh-keygen -Y sign -f test_sign_key -n git hola.txt
+### using test_sign_key in ../testdata
+
+msg_sig_pair_hola_mundo = (
+    b"""\
+Hola Mundo
+""",
+    b"""\
+-----BEGIN SSH SIGNATURE-----
+U1NIU0lHAAAAAQAAADMAAAALc3NoLWVkMjU1MTkAAAAgZz+kZMT7JBP0t1l1HQ0K8CduhZ
+XTBP/l3sXkZMqTtAkAAAADZ2l0AAAAAAAAAAZzaGE1MTIAAABTAAAAC3NzaC1lZDI1NTE5
+AAAAQDFSdQINV271MZ5VwFecGD8oJRob5Nb04r06oVVVCflwgbDLcezjmHJQo41/H3/HXj
+pQWO8AJXrx7gcPAcCFGQ0=
+-----END SSH SIGNATURE-----
+""",
+)
+
+crazy_ascii = b"Nobody expects the Spanish ..."
+crazy_unicode = "Nobody expects 🥘💃🐂 ...".encode()
+
+
+class CLITests(TestCase):
```

Comment:
Please add tests for the actual lower level functions - I'm not expecting us to actually provide the cli to users, they can use ssh-keygen itself for that.

Compatibility tests (such as for ssh-keygen here) should probably live in tests/compat, and run when ssh-keygen is actually installed on the system - consistently rather than just when a flag is set.

Reply:
Nice! I had not looked at tests/compat. Looks like a good approach. I'm going to think about what should go in tests/compat and what stays in tests/test_????.py.

I was imagining that, similar to `python3 -m dulwich ...`, an new possibility would be
```
python3 -m dulwich.ssh_keygen -Y ...
```
mainly for troubleshooting and debugging reasons. But I can't really think of a compelling reason for this versus somebody just cloning the Dulwich GitHub repo and running a `ssh_keygen.py` testing/compat/util script that does not ship with the Dulwich package.

I'll look into having `ssh_keygen.py` be a testing/troubleshootin/debugging script that lives in the Dulwich repo but not in the code that gets packaged.

Then maybe `sshsig.py` is a good filename for the Python script that does get packaged. Or `sshsiglib.py`. Any thoughts or prefs?

Reply:
What's the purpose of having ssh_keygen.py over just running the ssh-keygen command? I'd rather have proper unit tests for the lower level code than to have to run another command and verify that it behaves the same as ssh-keygen.

Reply:
Next commit is going to have all the command line option parsing cut out. I still have it working over at gitlab `dulwich-sshsiglib/sshsiglib` so I can add a utility back into dulwich later and separately if you'd like. You'll find the new commit to dulwich much more proper.

As far as reasoning behind having the CLI... the CLI is the unambiguous "ground truth", more so than the best most proper unit test. If some input is not working as expected with the low level code, a sensible troubleshooting step is 1st) see how it behaves via the CLI with the real ssh-keygen, then 2nd) confirm the ssh_keygen.py CLI matches or not. Otherwise there is ambiguity as to whether one is simply calling the low-level function improperly or in the wrong state etc... vs the low-level functions are actually not working properly.